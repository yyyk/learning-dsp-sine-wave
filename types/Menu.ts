export interface MenuLink {
  title: string
  path: string
}
