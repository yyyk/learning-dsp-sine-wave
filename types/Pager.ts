export interface PageLink {
  path: string
  title?: string
  slug?: string
}
