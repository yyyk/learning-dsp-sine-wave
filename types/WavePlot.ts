export interface Sine {
  level: number
  ratio: number
  coarse: number
  fine: number
  frequencyOffset: number
}
