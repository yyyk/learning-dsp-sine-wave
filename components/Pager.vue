<template>
  <div class="pager">
    <nuxt-link
      v-if="prev"
      :to="prev.path === '/index' ? '/' : prev.path"
      exact
      prefetch
      class="prev"
    >
      &#60;
      <span class="link-text">To the previous chapter</span>
    </nuxt-link>
    <nuxt-link v-if="next" :to="next.path" exact prefetch class="next">
      <span class="link-text">To the next chapter</span> &#62;
    </nuxt-link>
  </div>
</template>

<script lang="ts">
import Vue, { PropOptions } from 'vue'
import { PageLink } from '@/types/Pager'

export default Vue.extend({
  props: {
    prev: {
      type: Object,
      default: null
    } as PropOptions<PageLink | null>,
    next: {
      type: Object,
      default: null
    } as PropOptions<PageLink | null>
  }
})
</script>

<style lang="scss" scoped>
.pager {
  width: 100%;
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: flex-start;
  padding: 16px;

  & + .pager {
    margin-top: 16px;
  }

  .prev,
  .next {
    text-decoration: none;

    &:hover,
    &:focus,
    &:active {
      .link-text {
        text-decoration: underline;
      }
    }
  }

  .next {
    margin-left: auto;
  }

  // TODO: hide in mobile
  // .link-text {
  // }
}
</style>
