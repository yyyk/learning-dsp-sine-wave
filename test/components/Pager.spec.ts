import { mount, createLocalVue, RouterLinkStub } from '@vue/test-utils'
import Pager from '@/components/Pager.vue'
import { PageLink } from '@/types/Pager'

const localVue = createLocalVue()

describe('Pager.vue', () => {
  const mountFunction = (options = {}) => {
    return mount(Pager, {
      localVue,
      stubs: { NuxtLink: RouterLinkStub },
      ...options
    })
  }

  it('renders correctly', () => {
    const wrapper = mountFunction()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('exists', () => {
    const wrapper = mountFunction()
    expect(wrapper.exists()).toBe(true)
  })

  it('should have pager css class', () => {
    const wrapper = mountFunction()
    const parentDiv = wrapper.find('.pager')
    expect(parentDiv.exists()).toBe(true)
  })

  describe('Props', () => {
    describe("'prev' and 'next' are null", () => {
      const wrapper = mountFunction({
        propsData: {
          prev: null,
          next: null
        }
      })
      it("shouldn't have prev link", () => {
        const prevLink = wrapper.find('.prev')
        expect(prevLink.exists()).toBe(false)
      })
      it("shouldn't have next link", () => {
        const nextLink = wrapper.find('.next')
        expect(nextLink.exists()).toBe(false)
      })
    })

    describe("'prev' and 'next' are provided", () => {
      const wrapper = mountFunction({
        propsData: {
          prev: {
            path: '/'
          } as PageLink,
          next: {
            path: '/'
          } as PageLink
        }
      })
      it('should have prev link', () => {
        const prevLink = wrapper.find('.prev')
        expect(prevLink.exists()).toBe(true)
      })
      it('should have next link', () => {
        const nextLink = wrapper.find('.next')
        expect(nextLink.exists()).toBe(true)
      })
    })

    describe("'prev' is only provided", () => {
      const wrapper = mountFunction({
        propsData: {
          prev: {
            path: '/'
          } as PageLink,
          next: null
        }
      })
      it('should have prev link', () => {
        const prevLink = wrapper.find('.prev')
        expect(prevLink.exists()).toBe(true)
      })
      it("shouldn't have next link", () => {
        const nextLink = wrapper.find('.next')
        expect(nextLink.exists()).toBe(false)
      })
    })

    describe("'next' is only provided", () => {
      const wrapper = mountFunction({
        propsData: {
          prev: null,
          next: {
            path: '/'
          } as PageLink
        }
      })
      it("shouldn't have prev link", () => {
        const prevLink = wrapper.find('.prev')
        expect(prevLink.exists()).toBe(false)
      })
      it('should have next link', () => {
        const nextLink = wrapper.find('.next')
        expect(nextLink.exists()).toBe(true)
      })
    })
  })
})
