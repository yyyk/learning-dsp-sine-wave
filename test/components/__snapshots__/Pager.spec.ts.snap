// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pager.vue renders correctly 1`] = `
"<div class=\\"pager\\">
  <!---->
  <!---->
</div>"
`;
